internal/controller/connector_controller.go

```diff
@@ -0,0 +1,338 @@
+	log = logf.FromContext(ctx)
+
+	var connector appsv1beta1.Connector
+	if err := r.Get(ctx, req.NamespacedName, &connector); err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.Error(err, "fail to fetch Connector")
+		}
+
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	configMapNames := []string{}
+	for _, stream := range connector.Spec.Streams {
+		configMapNames = append(configMapNames, appsv1beta1.GenConnectorConfigMapName(connector.Spec.TemplateName, false)+"-for-"+stream)
+	}
+	var configs []map[string]interface{}
+	if connector.ObjectMeta.DeletionTimestamp.IsZero() {
+		cfgs, err := r.mergePatchesIntoConfigs(ctx, connector)
+		if err != nil {
+			log.Error(err, "fail to merge patches into config")
+
+			return ctrl.Result{}, err
+		}
+		configs = cfgs
+	}
+
+	for index, configMapName := range configMapNames {
+		var connectorConfigMap corev1.ConfigMap
+		if err := r.Get(ctx, types.NamespacedName{
+			Namespace: connector.Namespace,
+			Name:      configMapName,
+		}, &connectorConfigMap); err != nil {
+			if apierrors.IsNotFound(err) {
+				configWithService := map[string]interface{}{
+					"connector": configs[index],
+					"hstream": map[string]string{
+						"serviceUrl": "hstream://" + connector.Spec.HServerEndpoint,
+					},
+				}
+
+				configJson, _ := json.Marshal(configWithService)
+
+				err = r.Create(ctx, &corev1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: connector.Namespace,
+						Name:      configMapName,
+					},
+					Data: map[string]string{
+						"config.json": string(configJson),
+					},
+				})
+				if err != nil {
+					log.Error(err, "fail to create ConfigMap for Connector",
+						"Connector", connector.Name,
+						"ConfigMap", configMapName,
+					)
+
+					return ctrl.Result{}, err
+				}
+
+				err = r.createConnectorDeployment(ctx, connector, connector.Spec.Streams[index], configMapName)
+				if err != nil {
+					log.Error(err, "fail to create Deployment for Connector",
+						"Connector", connector.Name,
+						"Deployment", appsv1beta1.GenConnectorDeploymentName(connector.Name, connector.Spec.Streams[index]),
+					)
+
+					return ctrl.Result{}, err
+				}
+			} else {
+				log.Error(err, "fail to fetch ConfigMap which generated by Connector",
+					"Connector", connector.Name,
+					"ConfigMap", configMapName,
+				)
+
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	if connector.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&connector, connectorfinalizerName) {
+			controllerutil.AddFinalizer(&connector, connectorfinalizerName)
+
+			if err := r.Update(ctx, &connector); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(&connector, connectorfinalizerName) {
+			// Clear generated ConfigMaps if Connector is being deleted.
+			for _, configMapName := range configMapNames {
+				if err := appsv1beta1.DeleteAssociatedConfigMap(ctx, r.Client, types.NamespacedName{
```

Comment:
I think didn't need delete resource in our code, we just set `metadata.OwnerReferences` in children resource, and when they parents resource deleted, they will be delete.

Reply:
Refactored in [c1e4d16](https://github.com/hstreamdb/hstream-operator/pull/59/commits/c1e4d162cd13df954be6ca0076d1251814c48261). Thanks for pointing out the concept!